### PR TITLE
crypt: add pbkdf options for LUKS container manipulation

### PR DESCRIPTION
**Pull request by kubiko**
add support for additional pbkd options to be passed for operations on LUKS container.

### CODE REVIEW

File: crypt.go [hunk 1]
```diff
@@ -725,6 +740,24 @@ func InitializeLUKS2Container(devicePath, label string, key []byte, options *Ini
 				"--luks2-keyslots-size", fmt.Sprintf("%dk", options.KeyslotsAreaKiBSize))
 		}
 	}
+	if pbkdfOptions != nil {
+		if pbkdfOptions.PbkdfMmemory != 0 {
+			args = append(args,
+				"--pbkdf-memory", fmt.Sprintf("%d", pbkdfOptions.PbkdfMmemory))
```
Comment (pedronis) [inline crypt.go:746]: instead of Sprintf this can use strconv.Itoa

File: crypt.go [hunk 2]
```diff
@@ -643,6 +643,21 @@ type InitializeLUKS2ContainerOptions struct {
 	KeyslotsAreaKiBSize int
 }
 
+// AddRecoveryKeyToLUKS2ContainerOptions carries options for recovery key to LUKS2
+// containers.
+type PbkdfLUKS2ContainerOptions struct {
```
Comment (pedronis) [inline crypt.go:648]: by naming conventions this should be PBKDFLUKS2...

File: crypt.go [hunk 3]
```diff
@@ -725,6 +740,24 @@ func InitializeLUKS2Container(devicePath, label string, key []byte, options *Ini
...
+			args = append(args,
+				"--pbkdf-memory", fmt.Sprintf("%d", pbkdfOptions.PbkdfMmemory))
+		}
+		if pbkdfOptions.PkdfForceIterations != 0 {
+			args = append(args,
+				"--pbkdf-force-iterations", fmt.Sprintf("%d", pbkdfOptions.PkdfForceIterations))
+		} else {
+			args = append(args, "--iter-time", "100")
+		}
+		if pbkdfOptions.PbkdfParallel != 0 {
```
Comment (pedronis) [inline crypt.go:754]: I'm not entirely sure why this is not under the if/else ?  I let @chrisccoulson comment/double check on this

File: crypt.go [hunk 4]
```diff
@@ -725,6 +740,24 @@ func InitializeLUKS2Container(devicePath, label string, key []byte, options *Ini
 				"--luks2-keyslots-size", fmt.Sprintf("%dk", options.KeyslotsAreaKiBSize))
 		}
 	}
+	if pbkdfOptions != nil {
```
Comment (pedronis) [inline crypt.go:743]: this logic here and repeated in the other functions could be  extracted as a method on *PbkdfLUKS2ContainerOptions

```
func (o *PbkdfLUKS2ContainerOptions) appendToArgs(args []string, fallbackIterTimeMs int) []string {
   ...
}
```

File: crypt.go [hunk 5]
```diff
@@ -643,6 +643,21 @@ type InitializeLUKS2ContainerOptions struct {
...
+	// PbkdfMmemory set the memory cost for PBKDF (for Argon2i/id the number
+	// represents kilobytes).  Note that it is maximal value,
+	// PBKDF benchmark or available physical memory can decrease it.
+	PbkdfMmemory int
+	// Avoid PBKDF benchmark and set time cost (iterations) directly.
+	PkdfForceIterations int
+	// Set the parallel cost for PBKDF (number of threads, up to 4).
+	// Note that it is maximal value, it is decreased
+	// automatically if CPU online count is lower.
+	PbkdfParallel int
```
Comment (pedronis) [inline crypt.go:658]: as it's already in the struct name we can leave out the Pkdf prefixes here

**Review by pedronis — COMMENTED** [review]
did a pass

File: crypt.go [hunk 6]
```diff
@@ -795,10 +828,33 @@ func addKeyToLUKS2Container(devicePath string, existingKey, key []byte, extraOpt
...
-func AddRecoveryKeyToLUKS2Container(devicePath string, key []byte, recoveryKey RecoveryKey) error {
-	return addKeyToLUKS2Container(devicePath, key, recoveryKey[:], []string{
-		// use argon2i as the KDF with an increased cost
-		"--pbkdf", "argon2i", "--iter-time", "5000"})
+func AddRecoveryKeyToLUKS2Container(devicePath string, key []byte, recoveryKey RecoveryKey, pbkdfOptions *PbkdfLUKS2ContainerOptions) error {
+	args := []string{
+		// use argon2i as the KDF with reduced cost. This is done because the supplied input key has an
+		// entropy of at least 32 bytes, and increased cost doesn't provide a security benefit because
+		// this key and these settings are already more secure than the recovery key. Increased cost
+		// here only slows down unlocking.
```
Comment (chrisccoulson) [inline crypt.go:836]: I don't think replacing the original comment was intentional.

File: crypt.go (hunk 5 above)
Comment (chrisccoulson) [inline crypt.go:658]: Is an option to configure the parallelism necessary?

**Review by chrisccoulson — COMMENTED** [review]
Thanks for this - I've left a few comments inline. Were you planning to add some tests for this?

File: crypt.go [hunk 7]
```diff
@@ -725,6 +740,24 @@ func InitializeLUKS2Container(devicePath, label string, key []byte, options *Ini
...
+	if pbkdfOptions != nil {
+		if pbkdfOptions.PbkdfMmemory != 0 {
+			args = append(args,
+				"--pbkdf-memory", fmt.Sprintf("%d", pbkdfOptions.PbkdfMmemory))
+		}
+		if pbkdfOptions.PkdfForceIterations != 0 {
+			args = append(args,
+				"--pbkdf-force-iterations", fmt.Sprintf("%d", pbkdfOptions.PkdfForceIterations))
+		} else {
+			args = append(args, "--iter-time", "100")
```
Comment (pedronis) [inline crypt.go:752]: we should also make iter-itme itself an option (and use the defaults if it's unset as well)

File: crypt.go [hunk 8]
```diff
@@ -795,10 +828,33 @@ func addKeyToLUKS2Container(devicePath string, existingKey, key []byte, extraOpt
...
+		}
+		if pbkdfOptions.PbkdfParallel != 0 {
+			args = append(args,
+				"--pbkdf-parallel", fmt.Sprintf("%d", pbkdfOptions.PbkdfParallel))
+		}
+	} else {
+		args = append(args, "--iter-time", "5000")
```
Comment (pedronis) [inline crypt.go:855]: same

**Review by pedronis — COMMENTED** [review]
we should also let specify the actual iter-time possibly from outside

File: crypt.go (hunk 7 above)
Comment (pedronis) [reply crypt.go:752]: for this one maybe is not so important

File: crypt.go (hunk 8 above)
Comment (pedronis) [reply crypt.go:855]: here is useful though

**Pull request merged**